### DISCUSSION TITLE
Issue with Disk Charts: Reserved Space for Root Showing Incorrect Value for One Partition

### DISCUSSION THREAD

**Discussion by ghost** [netdata/netdata#17812] [Q&A]
**Summary:**
I have created my ext4 file systems without reserved space for root (reserved block count) during the Debian12 installation using the `-m 0` option with `mkfs.ext4`. This configuration is correctly applied to all partitions, and I am able to see the 0% reserved space for all my partitions

#### **Partition List:**
```bash
root@DEB-ZBOXMI35-GXXXX472XXXXX:~# lsblk 
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
nvme0n1     259:0    0 232.9G  0 disk 
├─nvme0n1p1 259:1    0   512M  0 part /boot/efi
├─nvme0n1p2 259:2    0    18G  0 part /
├─nvme0n1p3 259:3    0     6G  0 part /opt/debian12/log
└─nvme0n1p4 259:4    0 208.4G  0 part /var/lib/docker
```

**Verification of Reserved Block Count:**
All partitions show a reserved block count of 0:

```bash
root@DEB-ZBOXMI35-GXXXX472XXXXX:~# sudo dumpe2fs -h /dev/nvme0n1p2 | grep -i "reserved block count"
dumpe2fs 1.47.0 (5-Feb-2023)
Reserved block count:     0

root@DEB-ZBOXMI35-GXXXX472XXXXX:~# sudo dumpe2fs -h /dev/nvme0n1p3 | grep -i "reserved block count"
dumpe2fs 1.47.0 (5-Feb-2023)
Reserved block count:     0

root@DEB-ZBOXMI35-GXXXX472XXXXX:~# sudo dumpe2fs -h /dev/nvme0n1p4 | grep -i "reserved block count"
dumpe2fs 1.47.0 (5-Feb-2023)
Reserved block count:     0
```

#### **Netdata Configuration:**

Here is my current Netdata configuration for disk space:

```ini
[plugin:proc:diskspace]
    update every = 60
    exclude space metrics on paths = /proc/* /sys/* /var/run/user/* /run/user/* /var/lib/docker/* /dev /dev/shm /run /run/* 
    inodes usage for all disks = no

[plugin:proc:/proc/diskstats]
    exclude disks = loop* ram* boot rootfs rex-kmshare ext4
    merged operations for all disks = no
    backlog for all disks = no
    i/o time for all disks = no
    enable new disks detected at runtime = no

[plugin:proc:/proc/diskstats:sda]
    enable = yes
    enable performance metrics = auto
    bandwidth = auto
    operations = auto
    merged operations = no
    queued operations = auto
    utilization percentage = auto
    extended operations = no

[plugin:proc:/proc/diskstats:nvme0n1]
    enable = yes
    enable performance metrics = auto
    bandwidth = auto
    operations = auto
    merged operations = no
    queued operations = auto
    utilization percentage = auto
    extended operations = no
```

In Netdata, the reserved space for root is correctly shown as 0% for 2 partitions ( /root & /var/lib/docker) out of 3 partitions.

![2024_06_04_0vt_Kleki](https://github.com/netdata/netdata/assets/87641643/e7de3169-14b8-4aba-8f27-771bcce873fa)

![2024_06_04_0vv_Kleki](https://github.com/netdata/netdata/assets/87641643/21a55ff1-6012-4700-b537-f2a9f514edf4)

### Issue:
But the reserved space for root is incorrectly displays as 0.2 for one partition ( /opt/debian12/log) but the Internally its value is set to 0 only.
![2024_06_04_0wf_Kleki](https://github.com/netdata/netdata/assets/87641643/eda566dd-d1b3-462a-858d-465d05027449)

5. **Manual Calculation of Reserved Space:**
   Manually calculated reserved space to cross-verify the displayed value:
```
root@DEB-ZBOXMI35-GXXXX472XXXXX:~# lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
nvme0n1     259:0    0 232.9G  0 disk 
├─nvme0n1p1 259:1    0   512M  0 part /boot/efi
├─nvme0n1p2 259:2    0    18G  0 part /
├─nvme0n1p3 259:3    0     6G  0 part /opt/debian12/log
└─nvme0n1p4 259:4    0 208.4G  0 part /var/lib/docker

root@DEB-ZBOXMI35-GXXXX472XXXXX:~# sudo dumpe2fs -h /dev/nvme0n1p3 | grep -i "reserved block count"
dumpe2fs 1.47.0 (5-Feb-2023)
Reserved block count:     0
root@DEB-ZBOXMI35-GXXXX472XXXXX:~# 
```
   
Confirmed that the reserved block count is indeed 0.

**Request for Assistance:**
I am aware that we can hide this dimension from Netdata charts, but I don't want to hide it. I want to understand the underlying mechanism and why it is showing 0.2% for this partition despite the actual reserved 

Any guidance on resolving this discrepancy would be greatly appreciated.

Thank you in advance for your help!

**Comment by ilyam8** [comment]
>  want to understand the underlying mechanism and why it is showing 0.2% for this partition despite the actual reserved

Hi, @fti-sfuke. Netdata does [statvfs()](https://man7.org/linux/man-pages/man3/statvfs.3.html) for every [mount point to gather disk space metrics](https://github.com/netdata/netdata/blob/29f5199d5fd28afb443759042c4568c6333cbaf7/src/collectors/diskspace.plugin/plugin_diskspace.c#L454). 

See

https://github.com/netdata/netdata/blob/29f5199d5fd28afb443759042c4568c6333cbaf7/src/collectors/diskspace.plugin/plugin_diskspace.c#L181-L187


Used values:

| Value    | Description |
| -------- | ------- |
| `f_blocks`  | Size of fs in f_frsize units    |
| `f_bavail`  | Number of free blocks for unprivileged users    |
| `f_bfree`  | Number of free blocks    |

"reserved_for_root" calculation: `f_bfree` - `f_bavail`.
